src/test/java/org/wise/portal/dao/user/impl/HibernateUserDaoTest.java

```diff
@@ -17,186 +17,504 @@
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByLastName_WithExistingLastName_ShouldReturnUser() {
+    List<User> users = userDao.retrieveStudentsByLastName("Squarepants");
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByUsername_WithNonExistingUsername_ShouldNotReturnAnyUser() {
+    List<User> users = userDao.retrieveStudentsByUsername("BillD0101");
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByUsername_WithExistingUsername_ShouldReturnUser() {
+    List<User> users = userDao.retrieveStudentsByUsername("SpongebobS0101");
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByGender_WithNotUsedGender_ShouldNotReturnAnyUser() {
+    List<User> users = userDao.retrieveStudentsByGender("female");
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByGender_WithUsedGender_ShouldReturnUser() {
+    List<User> users = userDao.retrieveStudentsByGender("male");
+    assertEquals(2, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByNameAndBirthday_WithNonExistingBirthday_ShouldNotReturnAnyUser() {
+    List<User> users = userDao.retrieveStudentsByNameAndBirthday("Spongebob", "Squarepants", 1, 2);
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveStudentsByNameAndBirthday_WithExistingBirthday_ShouldReturnUser() {
+    List<User> users = userDao.retrieveStudentsByNameAndBirthday("Spongebob", "Squarepants", 1, 1);
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveTeachersByName_WithNonExistingName_ShouldNotReturnAnyUser() {
+    List<User> users = userDao.retrieveTeachersByName("Master", "Splinter");
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveTeachersByName_WithExistingName_ShouldReturnUser() {
+    List<User> users = userDao.retrieveTeachersByName("Mrs", "Puff");
+    assertEquals(1, users.size());
+  }
+
+  @Test
+  public void retrieveTeachersByResetPasswordKey_WithExistingResetPasswordKey_ShouldReturnUser() {
+    teacher1.getUserDetails().setResetPasswordKey("qwerty");
+    User user = userDao.retrieveByResetPasswordKey("qwerty");
+    assertEquals("MrsPuff", user.getUserDetails().getUsername());
+  }
+
+  @Test
+  public void retrieveTeacherUsersJoinedSinceYesterday_WithNewRecentTeachers_ShouldReturnUsers() {
+    List<User> users = userDao.retrieveTeacherUsersJoinedSinceYesterday();
+    assertEquals(2, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInSinceYesterday_NoRecentLogins_ShouldReturnNoUsers() {
+    List<User> users = userDao.retrieveTeacherUsersWhoLoggedInSinceYesterday();
+    assertEquals(0, users.size());
+  }
+
+  @Test
+  public void retrieveTeacherUsersWhoLoggedInSinceYesterday_WithRecentLogin_ShouldReturnUsers() {
+    ((TeacherUserDetails) teacher1.getUserDetails()).setLastLoginTime(new Date());
```

Comment:
```suggestion
    teacher1.getUserDetails().setLastLoginTime(new Date());
```
No need to cast because setLastLoginTime is a method in MutableUserDetails class. Change for all other instances.